titan/features/partner_tracing.py

```diff
@@ -21,35 +62,167 @@ def update_agent(self, model: "model.TITAN"):
+                    .sex_type[self.agent.sex_type]
+                    .drug_type[self.agent.drug_type]
+                    .haart
+                )
+                self.agent.haart.initiate(model.run_random, haart_params, "prob") #deterministic. They will enroll in haart.
+                self.re_initiated = True
+                self.re_initiated_time = model.time
+                
+  
         # stop tracing of this agent if time
         if self.active and model.time >= self.time + params.trace_duration:
             self.active = False
             self.time = None
+        
+        # this is to ensure agents are able to test negative multiple times throughout the simulation if contacted. On the otherhand, agents cannot test positive multiple times
+        if self.tested_negative and model.time > self.tested_negative_time:
+            self.tested_negative = False
+            self.tested_negative_time = None
+        
+        # this is to ensure agents are able to participate multiple times throughout the simulation 
+        if self.ps_participant and model.time > self.ps_participant_time:
+            self.ps_participant = False
+            self.ps_participant_time = None
+    
+    def trace_partners(self, params, model: "model.TITAN"):
+        for ptnr in self.agent.get_partners(params.bond_type):
+            
+            # step 3
+            self.info = True if model.run_random.random() < params.tracing_prob else False # can only trace if participant provides information
+            # step 4
+            self.contact = True if model.run_random.random() < params[ptnr.race].contact_prob else False # contact stochastically
+            
+            if self.info:
+                self.info_stat += 1
+            
+            if (
+                self.info
+                and self.contact
+            ):
+                self.contact_stat += 1
+                ptnr.partner_tracing.active = True  # type: ignore[attr-defined]
+                ptnr.partner_tracing.time = model.time  # type: ignore[attr-defined]
+                
+    def test_agent(self, agent_exposure, params, model: "model.TITAN"):
+        
+        # if agent has hiv, mark as diagnosed
+        if (
+            agent_exposure.active
+            and model.run_random.random() < params.dx_prob
+        ):    
+            agent_exposure.diagnose(model) # markes agent with positive hiv diagnoses (not stochastic)
+            self.ps_dx = True
+            self.ps_dx_time = model.time
+            self.add_positive_count()
+            
+        # if agent does not have hiv, mark as tested negative
+        else:
+            self.tested_negative = True
+            self.tested_negative_time = model.time
+            self.add_negative_count()
+        
+    def set_stats(self, stats: Dict[str, int], time: int):
+        
+        if self.participate:
+            if self.participate_time == time:
```

Comment:
looks like these two can be combined to `if self.participate and self.participate_time == time`